torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -519,6 +525,42 @@ def _root_pre_forward(
     _p_assert(state._is_root is not None, "Expects a root FSDP to have been set")
     if not state._is_root:
         return args, kwargs
+
+    # We cast buffers back to fp32 if we're forcing full precision. In addition, we check if buffers
+    # are in full precision and if we should cast them back to lower precision, which happens when
+    # exiting eval() mode.
+
+    cast_buffers_to_full_prec = False
+    for handle in state._handles:
+        if handle._force_full_precision:
+            cast_buffers_to_full_prec = True
+            break
+
+    if cast_buffers_to_full_prec:
+        _cast_buffers_to_dtype_and_device(
+            buffers=dict(module.named_buffers()).values(),
+            buffer_dtypes=list(state._buffer_name_to_orig_dtype.values()),
+            device=state.compute_device
+        )
+        # This flag is only set when we cast buffers to full precision, to avoid the
+        # CPU overhead that can stem from retrieving all buffers and their types in the
+        # following else branch.
+        state._needs_restore_check = True
+    elif hasattr(state, '_needs_restore_check') and state._needs_restore_check:
+        # Check if buffers are in full precision and we need to cast them
+        # back down.
+        buffers, buffer_dtypes_for_computation = _get_buffers_and_dtypes_for_computation(state, module)
+        # buffers, buffer_dtypes_for_computation = list(buffers), list(buffer_dtypes_for_computation)
+        if len(buffers) > 0 and len(buffer_dtypes_for_computation) > 0:
+            buffer, buffer_dtype_for_computation = buffers[0], buffer_dtypes_for_computation[0]
+            if buffer.dtype != buffer_dtype_for_computation:
+                # Assume we have to cast everything if there is one mismatch
```

Comment:
Why do we only check the 0th buffer for mismatch?

Reply:
good point, moving this to any() check.